rework micro app routing

MonPote opened the pull request:
**Context**: 

The goal of this PR is to prepare the ability to load several versions of the same micro app and be able to load micro app from a micro app.
To achieve that, we need to  : 
- Be able to rely only on the discoveryUrl to render the Route instead of mixing it with the value of the config.json.
- Rework the way we render the micro app's `Routes`
 (see: `InternalRouter` in `shell-ui/src/FederatedApp.tsx`)

In the previous implementation we use `useDiscoveredViews` to render the Routes and render the navbar entries.
This was convenient because we reuse the hooks for the rendering the `Navbar` (see: `useFederatedNavbarEntries`) and rendering the routes (DRY) but also cause some unnecessary coupling.
The route need to be exposed in the navbar in order to be accessible in the UI.

In a close futur, we want to be able to render some FederatedComponent without it to be exposed in the navbar.
For example, we might want to switch micro app with a selector instead of click on the navbar.

Here are some notable change made : 
- we keep `useDiscoveredViews` for the navbar entries
- we create `useFederatedRoutes` to compute the federatedRoutes
- we add the url from where the micro app has been fetch in the metadata (cf: `EnrichedBuildtimeWebFinger`)in order to be able to find the right micro app config since we can now have several time the same app.

Those changes simplify a lot the `FederatedApp.tsx` however we lost the capability to render only `Routes` which the user have access because this information is only present on the config.json. It means that we rely on the API to filter what the user can or not do (it's already the case). The Navbar will still render only the correct entries though.

I've also add a checking to warn the user when `appHistoryBasePath` is not correct.

bert-e commented:
# Hello monpote,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/metalk8s/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @xaf-scality

* @rdebay-scality

* @eg-ayoub

* @teddyandrieux

* @jbwatenbergscality

* @thomasdanan

* @chengyanjin

* @ezekiel-alexrod

* @vdaviot

* @m4nch0t

MonPote committed:
add checking on appHistoryBasePath

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @vdaviot

* @chengyanjin

* @ezekiel-alexrod

* @m4nch0t

* @rdebay-scality

* @xaf-scality

* @jbwatenbergscality

* @eg-ayoub

* @teddyandrieux

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @ezekiel-alexrod

* @teddyandrieux

* @m4nch0t

* @jbwatenbergscality

* @rdebay-scality

* @thomasdanan

* @eg-ayoub

* @xaf-scality

* @vdaviot

* @chengyanjin

MonPote committed:
update target tsconfig target to es2023 to be able to use toSorted

MonPote committed:
create useFederationsRoutes

The goal of this PR is to prepare the ability to load several versions of the same micro app
and be able to load micro app from a micro app.
To achieve that, we need to  :
- Be able to rely only on the discoveryUrl to render the Route instead of mixing it with the value of the config.json.
- Rework the way we render the micro app's `Routes`
 (see: `InternalRouter` in `shell-ui/src/FederatedApp.tsx`)

In the previous implementation we use `useDiscoveredViews` to render the Routes and render the navbar entries.
This was convenient because we reuse the hooks for the rendering the `Navbar` (see: `useFederatedNavbarEntries`)
and rendering the routes (DRY) but also cause some unnecessary coupling.
The route need to be exposed in the navbar in order to be accessible in the UI.

In a close futur, we want to be able to render some FederatedComponent without it to be exposed in the navbar.
For example, we might want to switch micro app with a selector instead of click on the navbar.

Here are some notable change made :
- we keep `useDiscoveredViews` for the navbar entries
- we create `useFederatedRoutes` to compute the federatedRoutes
- we add the url from where the micro app has been fetch in the metadata (cf: `EnrichedBuildtimeWebFinger`)
in order to be able to find the right micro app config since we can now have several time the same app.

Those changes simplify a lot the `FederatedApp.tsx` however we lost the capability to render only `Routes`
which the user have access. It means that we rely on the API to filter what the user can or not do (it's already the case).

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @ezekiel-alexrod

* @eg-ayoub

* @thomasdanan

* @vdaviot

* @rdebay-scality

* @chengyanjin

* @teddyandrieux

* @jbwatenbergscality

* @xaf-scality

* @m4nch0t

JBWatenbergScality commented on shell-ui/src/initFederation/ConfigurationProviders.tsx:
```diff
@@ -318,9 +341,52 @@ export function useDiscoveredViews(): ViewDefinition[] {
+export function useFederatedRoutes(): FederatedRoute[] {
+  const { retrieveConfiguration } = useConfigRetriever();
+  const { retrieveDeployedApps } = useDeployedAppsRetriever();
+  const deployedApps = retrieveDeployedApps();
+
+  deployedApps.forEach((app) => {
+    if (app.appHistoryBasePath.endsWith('/')) {
+      throw new Error(
+        `appHistoryBasePath of app ${app.name} ends with a /, this is not allowed`,
+      );
+    }
```
Do we really want to run this on each render of the component holding this hook ? I think we can at least move a good part of this hook content into a useMemo

JBWatenbergScality commented on shell-ui/src/initFederation/ConfigurationProviders.tsx:
```diff
@@ -318,9 +341,52 @@ export function useDiscoveredViews(): ViewDefinition[] {
+export function useFederatedRoutes(): FederatedRoute[] {
+  const { retrieveConfiguration } = useConfigRetriever();
+  const { retrieveDeployedApps } = useDeployedAppsRetriever();
+  const deployedApps = retrieveDeployedApps();
+
+  deployedApps.forEach((app) => {
+    if (app.appHistoryBasePath.endsWith('/')) {
+      throw new Error(
+        `appHistoryBasePath of app ${app.name} ends with a /, this is not allowed`,
+      );
+    }
+  });
+
+  const federatedRoutes = deployedApps.flatMap((app) => {
+    const appBuildConfig = retrieveConfiguration<'build'>({
+      configType: 'build',
+      name: app.name,
+      url: app.url,
+    });
+
+    if (!appBuildConfig) {
+      return [];
+    }
```
nit: we are iterating twice on deployedApps within the same scope, can we iterate only once on this collection?

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @m4nch0t

* @eg-ayoub

* @ezekiel-alexrod

* @vdaviot

* @jbwatenbergscality

* @chengyanjin

* @thomasdanan

* @xaf-scality

* @rdebay-scality

MonPote committed:
refactor(routing): optimize deployedApps iteration and memoize route computation

MonPote commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/131.0`



The following branches will **NOT** be impacted:

* `development/123.0`
* `development/124.0`
* `development/124.1`
* `development/125.0`
* `development/126.0`
* `development/127.0`
* `development/128.0`
* `development/129.0`
* `development/130.0`
* `development/2.0`
* `development/2.1`
* `development/2.10`
* `development/2.11`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/131.0`



The following branches have **NOT** changed:

* `development/123.0`
* `development/124.0`
* `development/124.1`
* `development/125.0`
* `development/126.0`
* `development/127.0`
* `development/128.0`
* `development/129.0`
* `development/130.0`
* `development/2.0`
* `development/2.1`
* `development/2.10`
* `development/2.11`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`



Please check the status of the associated issue RING-50866.

Goodbye monpote.
